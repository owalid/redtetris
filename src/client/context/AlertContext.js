import createDataContext from "./createDataContext";
export const ALERT_POP = 'ALERT_POP'

const initialState = {
  message: '',
  type: null,
};

const alertReducer = (state, action) => {
  switch(action.type){
    case ALERT_POP:
      return { message: action.message, type: action.typeAlert }
    default: 
      return state
  }
};

const sendAlert = (dispatch) => (message, typeAlert) =>
  dispatch({ type: ALERT_POP, typeAlert, message })

export const { Provider, Context } = createDataContext(
  //reducer :
  alertReducer,

  //action functions :
  {
    sendAlert
  },

  //initialState :
  {
   ...initialState
  }
);