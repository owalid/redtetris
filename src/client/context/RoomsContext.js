import createDataContext from "./createDataContext";

const initialState = {
  rooms: null
};

const RoomsReducer = (state, payload) => {
  switch (payload.type) {
    case "updateRooms":
      return { ...state, rooms: payload.rooms };
    default:
      return state;
  }
};

const updateRooms = dispatch => rooms => {
  if (Object.keys(rooms).includes("_data")) {
    dispatch({ type: "updateRooms", rooms: rooms._data });
  } else {
    dispatch({ type: "updateRooms", rooms: rooms });
  }
}

export const { Provider, Context } = createDataContext(
  //reducer :
  RoomsReducer,

  //action functions :
  {
    updateRooms
  },

  //initialState :
  {
   ...initialState
  }
);