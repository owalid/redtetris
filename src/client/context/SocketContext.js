import React, { createContext } from "react"
import socketIOClient from "socket.io-client"
require('dotenv').config() 

export const SocketContext = createContext();

export const SocketContextProvider = ({ children }) => {
  const socketClient = socketIOClient(process.env.API_URL);  
  const sendSocket = (type, data = false) => {
    socketClient.emit(type, data);
  }
  
  return (
    <SocketContext.Provider
      value={{ socketClient, sendSocket }}
    >
      { children }
    </SocketContext.Provider>
  )
}