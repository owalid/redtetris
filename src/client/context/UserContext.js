import createDataContext from "./createDataContext";

const initialState = {
  uuidRoom: null,
  player: null
};

const UserReducer = (state, payload) => {
  switch (payload.type) {
    case "updateUuidRoom": {
      return {  ...state, uuidRoom: payload.uuidRoom };
    }
    case "updatePlayer": 
      return {  ...state, player: payload.player };
    default:
      return state;
  }
};

const updateUuidRoom = dispatch => uuidRoom => 
  dispatch({ type: "updateUuidRoom", uuidRoom });

const updatePlayer = dispatch => player => 
  dispatch({ type: "updatePlayer", player });

export const { Provider, Context } = createDataContext(
  //reducer :
  UserReducer,

  //action functions :
  {
    updateUuidRoom,
    updatePlayer
  },

  //initialState :
  {
   ...initialState
  }
);