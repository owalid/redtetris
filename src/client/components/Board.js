import React, { useContext, useEffect } from 'react';
import _ from 'lodash';
import { Card } from '@material-ui/core';

import { Context as UserContext } from "../context/UserContext";
import { SocketContext } from "../context/SocketContext";

import Preview from './Preview'
import Game from './Game'
import Chat from './Chat';
import ReGame from './ReGame';
import VisitorView from './VisitorView';

const boxProps = {
  bgcolor: 'background.paper',
  bordercolor: 'text.primary',
  border: 1,
  borderradius: "borderRadius",
  style: { backgroundColor: 'white', height: '100%', width: '100%'},
  m: 1
};


const Board = ({ finalScore, song, currentRoom, isEnd, uuidRoom, mapGame, mapsGamePreview, isAlone, score, sheet }) => {
  const { state: { player } } = useContext(UserContext);
  const { sendSocket } = useContext(SocketContext);


  useEffect(() => {
    if (isEnd === true && currentRoom) {
      sendSocket('server/end-game-visitor', { channel: uuidRoom })
      sendSocket('server/end-game', { channel: uuidRoom, uuidUser: player.uuid })
    }
  }, [isEnd])

  if ((player && player.visitor)
      || (player && Object.keys(currentRoom).includes('players')
            && Object.keys(currentRoom.players).includes(player.uuid)
            && currentRoom.players[player.uuid]
            && currentRoom.players[player.uuid].visitor)) {

    return (
      <div className="test--visitor-player">
        <VisitorView
          uuidRoom={uuidRoom}
          boxProps={boxProps}
          isEnd={isEnd}
          currentRoom={currentRoom}
          player={player}
          mapsGamePreview={mapsGamePreview}
          isAlone={isAlone}
          />
      </div>
    )

  } else {

    return (
      <div className="d-flex jcnt--start aitems--fs fdir--row test--player">
    
        {
          (isEnd)
          ? (<ReGame
              player={player}
              currentRoom={currentRoom}
              finalScore={finalScore}
            />)
          : (<div className="width-100">
                <Card {...boxProps} variant="outlined">
                  <Game
                    mapGame={mapGame}
                    song={song}
                    isOtherUser={false}
                  />
                </Card>
              </div>)
        }

        <div className="aself--str">
          <Preview
            mapsGamePreview={mapsGamePreview}
            isVisitor={false}
            isAlone={isAlone}
            score={score}
            sheet={sheet}
            uuidRoom={uuidRoom}
          />
        </div>
        <div style={{ width: '50vw' }}>
          <Card {...boxProps} variant="outlined">
            <Chat uuidRoom={uuidRoom} />
          </Card>
        </div>
      </div>
    );

  }
}

export default Board
