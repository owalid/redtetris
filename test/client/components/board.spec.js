import React, { useContext, useEffect } from "react";
import _ from 'lodash'
import '@testing-library/jest-dom/extend-expect';
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Board from '../../../src/client/components/Board'
import { render } from '@testing-library/react'
import { Context as UserContext } from "../../../src/client/context/UserContext";
import { Context as RoomsContext } from "../../../src/client/context/RoomsContext";
import { TestAppUserProvider } from "../helpers/UserContext";
import { player_instance1, rooms_instance, visitor_player, room1, player_visitor_instance } from '../helpers/data'
import { TestAppSocketProvider } from "../helpers/SocketContext";
import { TestAppRoomsProvider } from "../helpers/RoomsContext";


Enzyme.configure({ adapter: new Adapter() });
describe('Board component', () => {
  it('Is exists', () => {
    const CurrentUserSetter = ({ children }) => {
      const { updateUuidRoom, updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updateUuidRoom("112233445566778899");
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const Wrapper = () => (
      <CurrentUserSetter>
        <Board />
      </CurrentUserSetter>
    )
    const wrapper = shallow(<Wrapper />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Can mount', () => {
    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Board
                  song={false}
                  currentRoom={room1}
                  isEnd={room1.players[player_instance1.uuid].end}
                  mapGame={room1.players[player_instance1.uuid].currentMapGame}
                  isAlone={Object.keys(room1.players).length === 1}
                  mapsGamePreview={_.filter(room1.players, item => item.uuid !== player_instance1.uuid && !item.visitor)}
                  score={room1.players[player_instance1.uuid].score}
                  sheet={room1.players[player_instance1.uuid].sheets[0]}
                  finalScore={room1.finalScore}
                  uuidRoom={room1.channel}
                />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )
    render(<Wr />);
  })

  it('Board Visitor', () => {
    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(player_visitor_instance);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Board
                  song={false}
                  currentRoom={room1}
                  isEnd={room1.players[player_visitor_instance.uuid].end}
                  mapGame={room1.players[player_visitor_instance.uuid].currentMapGame}
                  isAlone={Object.keys(room1.players).length === 1}
                  mapsGamePreview={_.filter(room1.players, item => item.uuid !== player_visitor_instance.uuid && !item.visitor)}
                  score={room1.players[player_visitor_instance.uuid].score}
                  sheet={room1.players[player_visitor_instance.uuid].sheets[0]}
                  finalScore={room1.finalScore}
                  uuidRoom={room1.channel}
                />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )
    const { container } = render(<Wr />);
    const container_not_visitor = container.querySelector('.test--visitor-player')
    expect(container_not_visitor).toBeNull()
  })

  it('Board Player', () => {
    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(player_instance1);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Board
                  song={false}
                  currentRoom={room1}
                  isEnd={room1.players[player_instance1.uuid].end}
                  mapGame={room1.players[player_instance1.uuid].currentMapGame}
                  isAlone={Object.keys(room1.players).length === 1}
                  mapsGamePreview={_.filter(room1.players, item => item.uuid !== player_instance1.uuid && !item.visitor)}
                  score={room1.players[player_instance1.uuid].score}
                  sheet={room1.players[player_instance1.uuid].sheets[0]}
                  finalScore={room1.finalScore}
                  uuidRoom={room1.channel}
                />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )
    const { container } = render(<Wr />);
    const container_not_visitor = container.querySelector('.test--player')
    expect(container_not_visitor).not.toBeNull()
  })

});
