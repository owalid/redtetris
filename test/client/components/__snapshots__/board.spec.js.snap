// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component Is exists 1`] = `
<CurrentUserSetter>
  <Board />
</CurrentUserSetter>
`;
